demo-multiroot-react-18/src/MultiRootEditorDemo.tsx

```diff
@@ -0,0 +1,286 @@
+		}
+	};
+
+	// Function to toggle read-only mode for selected root.
+	const toggleReadOnly = () => {
+		const root = editor!.model.document.selection.getFirstRange()!.root;
+
+		if ( !root || !root.rootName ) {
+			return;
+		}
+
+		const isReadOnly = disabledRoots.has( root.rootName );
+
+		if ( isReadOnly ) {
+			disabledRoots.delete( root.rootName );
+			editor!.enableRoot( root.rootName, SAMPLE_READ_ONLY_LOCK_ID );
+		} else {
+			disabledRoots.add( root.rootName );
+			editor!.disableRoot( root.rootName, SAMPLE_READ_ONLY_LOCK_ID );
+		}
+
+		setDisabledRoots( new Set( disabledRoots ) );
+	};
+
+	// Function to simulate an error in the editor.
+	// It is used to trigger watchdog to restart the editor.
+	const simulateError = () => {
+		setTimeout( () => {
+			const err: any = new Error( 'foo' );
+
+			err.context = editor;
+			err.is = () => true;
+
+			throw err;
+		} );
+	};
+
+	const addRoot = ( newRootAttributes: Record<string, unknown>, rootId?: string ) => {
+		const id = rootId || new Date().getTime();
+
+		for ( let i = 1; i <= numberOfRoots; i++ ) {
+			const rootName = `root-${ i }-${ id }`;
+
+			content[ rootName ] = '';
+			attributes[ rootName ] = { ...newRootAttributes, order: i * 10, row: id };
+		}
+
+		setContent( { ...content } );
+		setAttributes( { ...attributes } );
+		setNumberOfRoots( 1 );
+	};
+
+	const removeRoot = ( rootName: string ) => {
+		setContent( previousContent => {
+			const { [ rootName! ]: _, ...newContent } = previousContent;
+
+			return { ...newContent };
+		} );
+
+		setSelectedRoot( '' );
+	};
+
+	// Function to handle the addition of a new root element to the editor.
+	// This callback is necessary to correctly handle the multi root editor state.
+	const handleNewRoot = ( createRootElement: ( props?: Record<string, unknown> ) => JSX.Element ) => {
+		const element = createRootElement();
+
+		setElements( previousElements => [ ...previousElements, element ] );
+	};
+
+	// Function to handle the removal of a root element from the editor.
+	// This callback is necessary to correctly handle the multi root editor state.
+	const handleRemovedRoot = ( { rootName }: { rootName: string } ) => {
+		// Handling of undo operations.
+		if ( content[ rootName ] !== undefined ) {
+			removeRoot( rootName );
+		}
+
+		setElements( previousElements => previousElements.filter( element => element.props.id !== rootName ) );
+	};
+
+	// Grouping elements based on their row attribute and sorting them by order attribute.
+	// It is used only for presentation purposes.
+	const groupedElements = Object.entries(
+		elements
+			.sort( ( a, b ) => ( attributes[ a.props.id ].order as number ) - ( attributes[ b.props.id ].order as number ) )
+			.reduce( ( acc: Record<string, Array<JSX.Element>>, element ) => {
+				const row = attributes[ element.props.id ].row as string;
+				acc[ row ] = acc[ row ] || [];
+				acc[ row ].push( element );
+
+				return acc;
+			}, {} )
+	);
+
+	return (
+		<>
```

Comment:
Add comments saying which elements are optional, similar to other comments I've made.